snap/info_snap_yaml.go

```diff
@@ -369,52 +358,98 @@ func bindUnboundSlots(slotNames []string, snap *Info) error {
 	return nil
 }
 
-func convertToSlotOrPlugData(plugOrSlot, name string, data interface{}) (iface, label string, attrs map[string]interface{}, err error) {
+func convertToSlotOrPlugData(plugOrSlot, name string, data interface{}) (iface, label string, attrs map[string]interface{}, strattrs map[string]interface{}, err error) {
 	iface = name
 	switch data.(type) {
 	case string:
-		return data.(string), "", nil, nil
+		return data.(string), "", nil, nil, nil
 	case nil:
-		return name, "", nil, nil
+		return name, "", nil, nil, nil
 	case map[interface{}]interface{}:
 		for keyData, valueData := range data.(map[interface{}]interface{}) {
 			key, ok := keyData.(string)
 			if !ok {
 				err := fmt.Errorf("%s %q has attribute that is not a string (found %T)",
 					plugOrSlot, name, keyData)
-				return "", "", nil, err
+				return "", "", nil, nil, err
 			}
 			if strings.HasPrefix(key, "$") {
 				err := fmt.Errorf("%s %q uses reserved attribute %q", plugOrSlot, name, key)
-				return "", "", nil, err
+				return "", "", nil, nil, err
 			}
 			switch key {
 			case "interface":
 				value, ok := valueData.(string)
 				if !ok {
 					err := fmt.Errorf("interface name on %s %q is not a string (found %T)",
 						plugOrSlot, name, valueData)
-					return "", "", nil, err
+					return "", "", nil, nil, err
 				}
 				iface = value
 			case "label":
 				value, ok := valueData.(string)
 				if !ok {
 					err := fmt.Errorf("label of %s %q is not a string (found %T)",
 						plugOrSlot, name, valueData)
-					return "", "", nil, err
+					return "", "", nil, nil, err
 				}
 				label = value
 			default:
 				if attrs == nil {
 					attrs = make(map[string]interface{})
+					strattrs = make(map[string]interface{})
+				}
+				strval, err := validateAttr(valueData)
+				if err != nil {
+					return "", "", nil, nil, fmt.Errorf("attribute of %s %q: %v", plugOrSlot, name, err)
 				}
 				attrs[key] = valueData
+				strattrs[key] = strval
 			}
 		}
-		return iface, label, attrs, nil
+		return iface, label, attrs, strattrs, nil
 	default:
 		err := fmt.Errorf("%s %q has malformed definition (found %T)", plugOrSlot, name, data)
-		return "", "", nil, err
+		return "", "", nil, nil, err
+	}
+}
+
+// validateAttr validates an attribute value and returns a version of it with scalars canonically turned into strings.
+func validateAttr(v interface{}) (interface{}, error) {
+	switch x := v.(type) {
+	case string:
+		return x, nil
+	case bool:
+		return strconv.FormatBool(x), nil
+	case int:
+		return strconv.Itoa(x), nil
```

Comment:
Hmm.. but that's basically doing the same thing we're doing now, right?  For it to be effective, it'd have to do the trick we discussed back then, loading the strings in the original representation used in the file, so that we're validating the real input.

It feels like I'm missing something though. Apologies in advance.